### PR TITLE
Visitor fills out Membership Application

### PR DESCRIPTION
**Pull request by tochman**
PT Story: https://www.pivotaltracker.com/story/show/133940725

_As a visitor
In order to get a membership with SHF (which makes my business more valuable )
I need to be able to submit a Membership Application_ 

Changes proposed in this pull request:
1. creates a MembershipApplication class
2. adds MembershipsController
3. adds view to create and submit an application
4. replaces gem `haml` with `haml-rails` 
5. css for telephone and number fields
6. css, js and helper for flashes

skarmavbild 2016-11-11 kl 12 44 05
skarmavbild 2016-11-11 kl 12 45 40

### CODE REVIEW

**Comment by tochman** [comment]
I will not have time to work on this PR today or this weekend. If anybody want to continue working on this (we need to add some error handling and cover that in tests AND add css to the template used) here is what you need to do:
1. You need to be added to a relevant AgileVentures team on GitHub
2. You need to add me (tochman) as a remote 
3. You need to checkout my feature branch
   
   ```
   $ git checkout -t tochman/new_aplication
   ```
4. Add code and tests
5. Make commits and push up to MY remote
   
   ```
   $ git push tochman new_aplication
   ```

That will update the PR with your changes. Good luck.
I'll be back after the weekend but situation at work will make it impossible for me to contribute any code.

File: db/schema.rb [hunk 1]
```diff
@@ -10,9 +10,43 @@
...
+    t.datetime "updated_at",                          null: false
+    t.string   "first_name",             default: "", null: false
+    t.string   "last_name",              default: "", null: false
+    t.string   "street",                 default: "", null: false
+    t.integer  "postal_code",                         null: false
+    t.string   "city",                   default: "", null: false
+    t.integer  "phone"
+    t.bigint   "business_number"
+    t.index ["email"], name: "index_users_on_email", unique: true, using: :btree
+    t.index ["reset_password_token"], name: "index_users_on_reset_password_token", unique: true, using: :btree
```
Comment (tochman) [inline db/schema.rb]: Why is this here?

File: features/create_membership_application.feature [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    And I fill in "Company Name" with "Craft Academy"
+    And I fill in "Company Number" with "1234561234"
+    And I fill in "Contact Person" with "Thomas"
+    And I fill in "Company Email" with "info@craft.se"
+    And I fill in "Phone Number" with "031-1234567"
+    And I click on "Submit"
+    Then I should be on the landing page
+    And I should see "Thank you, Your application has been submitted"
+
+  Scenario Outline: Apply for membership sad path
```
Comment (tochman) [inline features/create_membership_application.feature]: I suggest renaming this to a more descriptive title. Avoid using `sad path`. It can be easily misunderstood by non technical stakeholders.

File: db/schema.rb (hunk 1 above)
Comment (thesuss) [reply db/schema.rb]: I didnt do db:drop. Will fix.

File: features/create_membership_application.feature (hunk 2 above)
Comment (thesuss) [reply features/create_membership_application.feature]: True, will do.

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -1,2 +1,8 @@
 module ApplicationHelper
+  def flash_class(level)
+    case level.to_sym
+      when :notice then 'success'
+      when :alert then 'danger'
+    end
+  end
```
Comment (tochman) [inline app/helpers/application_helper.rb:7]: Helpers needs to be covered in specs IMO

File: app/views/layouts/application.html.haml [hunk 4]
```diff
@@ -10,8 +10,9 @@
     = javascript_include_tag 'application', 'data-turbolinks-track': 'reload'
   %body.page.page-template.page-template-page-sidebar-none
     = render 'partials/top'
-    - flash.each do |key, value|
-      = value
+    .flashes
+      - flash.each do |key, value|
+        %div{:class => "alert alert-#{flash_class(key)} center"}= value
```
Comment (tochman) [inline app/views/layouts/application.html.haml]: Old hash synax.

File: app/helpers/application_helper.rb (hunk 3 above)
Comment (thesuss) [reply app/helpers/application_helper.rb:7]: I know, trying to find out how that is done, unless someone else beats me to it..

Comment (tochman) [reply app/helpers/application_helper.rb:7]: Take a look at: https://www.relishapp.com/rspec/rspec-rails/v/3-5/docs/helper-specs/helper-spec

Basically you want to test if the right string is returned by the helper when you pass in a value to it.

Comment (tochman) [reply app/helpers/application_helper.rb:7]: Something like this perhaps? 

```
RSpec.describe ApplicationHelper, type: :helper do
  describe '#flash_class' do

     it 'returns \'success\' on \'notice\'' do
       expect(helper.flash_class('notice')).to eq 'success'
     end

     it 'returns \'alert\' on \'danger\'' do
       expect(helper.flash_class('alert')).to eq 'denger'
     end
   end
end
```

Comment (thesuss) [reply app/helpers/application_helper.rb:7]: I think I solved it.

Comment (thesuss) [reply app/helpers/application_helper.rb:7]: Before I saw this comment..

File: spec/helpers/helper_spec.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe ApplicationHelper, type: :helper do
+  describe 'flashes' do
```
Comment (tochman) [inline spec/helpers/helper_spec.rb]: IMO you **should** reference the helper method you are testing. `#flash_class`

**Comment by tochman** [comment]
@thesuss Please update the PR description with your additions. Is there a lot of UI changes? Perhaps a screen shot?

File: app/controllers/memberships_controller.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    @membership = MembershipApplication.new
+  end
+
+  def create
+    @membership = MembershipApplication.new(membership_params)
+    if @membership.save
+      flash[:notice] = 'Thank you, Your application has been submitted'
+      redirect_to root_path
+    else
+      render new_membership_path
```
Comment (tochman) [inline app/controllers/memberships_controller.rb]: Are you sure this is working @thesuss?  I was under the impression that we `render` an action (like `render :new`) and `redirect_to` a path (like `redirect_to new_membership_path`). I haven't seen render used like this before. Interesting. 

Nonetheless, I would set this to  `render :new`

File: app/views/memberships/new.html.haml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    = f.label :contact_person, 'Contact Person'
+    = f.text_field :contact_person, class: 'wpcf7-form-control'
+
+    = f.label :company_email, 'Company Email'
+    = f.email_field :company_email, class: 'wpcf7-form-control'
+
+    = f.label :phone_number, 'Phone Number'
+    = f.telephone_field :phone_number, class: 'wpcf7-form-control'
+
+    = f.submit 'Submit'
```
Comment (AmberWilkie) [inline app/views/memberships/new.html.haml:26]: Shouldn't these form fields be in Swedish, if the site is in Swedish?

File: app/views/partials/_top.html.haml [hunk 8]
```diff
@@ -20,6 +20,8 @@
               = link_to 'Om', 'http://sverigeshundforetagare.se/'
             %li.page_item.page-item-2
               = link_to 'Mer', 'http://sverigeshundforetagare.se/'
+            %li.page_item.page-item-2
+              = link_to 'Apply for membership', new_membership_path
     / Social profiles
     .social-nav
       .tb
```
Comment (AmberWilkie) [inline app/views/partials/_top.html.haml:27]: Same here. I know I haven't been active - do you have a strategy for language?

File: features/step_definitions/basic_steps.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+end
+
+And(/^I fill in "([^"]*)" with "([^"]*)"$/) do |field, value|
+  fill_in field, with: value
+end
+
+When(/^I fill in the form with data :$/) do |table|
+  data = table.hashes.first
+  data.each do |label, value|
+    if !value.empty?
```
Comment (AmberWilkie) [inline features/step_definitions/basic_steps.rb]: `unless value.empty?` ?

Comment (thesuss) [reply features/step_definitions/basic_steps.rb]: Hehe, I took this from garden club :) Will change

File: app/views/memberships/new.html.haml (hunk 7 above)
Comment (thesuss) [reply app/views/memberships/new.html.haml:26]: I don't know what is best, translate as we go, or have one chore/feature for translating?

Comment (tochman) [reply app/views/memberships/new.html.haml:26]: One strategy is to add i18n and translation files (yaml).

Comment (thesuss) [reply app/views/memberships/new.html.haml:26]: I think we do it later, so it isn't a blocker for this one. Whatever decision is taken.

File: app/views/partials/_top.html.haml (hunk 8 above)
Comment (thesuss) [reply app/views/partials/_top.html.haml:27]: We will have one, but this PR can be as it is in regards to language.

**Review by diraulo — APPROVED** [review]
Looks good to me...
